consul.go

```diff
@@ -63,9 +63,16 @@ func watchConsulService(ctx context.Context, s servicer, tgt target, out chan<-
 				},
 			)
 			if err != nil {
-				grpclog.Errorf("[Consul resolver] Couldn't fetch endpoints. target={%s}; error={%v}", tgt.String(), err)
-				time.Sleep(bck.Duration())
-				continue
+				// No need to continue if the context is done/cancelled.
+				// We check that here directly because the check for the closed quit channel
+				// at the end of the loop is not reached when calling continue here.
+				if ctx.Err() != nil {
+					return
+				} else {
+					grpclog.Errorf("[Consul resolver] Couldn't fetch endpoints. target={%s}; error={%v}", tgt.String(), err)
+					time.Sleep(bck.Duration())
+					continue
+				}
```

Comment:
An alternative here could be to check for the `quit` channel while also providing a `default` case to not block in case the quit channel is not closed yet. This would allow this goroutine to not require knowing about the context object because it just communicates via channel, which probably follows best practices a bit more closely.
Like:
```suggestion
				select {
				case <-quit:
					return
				default:
					grpclog.Errorf("[Consul resolver] Couldn't fetch endpoints. target={%s}; error={%v}", tgt.String(), err)
					time.Sleep(bck.Duration())
					continue
				}
```